[DTensor] Moved `Transformer` sharding to staticmethod

awgu opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #121680
* __->__ #121660
* #121357
* #121360

To support FSDP + TP/SP unit tests, let us factor out the canonical TP/SP sharding of `Transformer` to a staticmethod that can be called by other unit tests.


Test Plan:
```
pytest test/distributed/tensor/parallel/test_tp_examples.py -k test_transformer_training
```

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @chauhang

Commit:
[DTensor] Moved `Transformer` sharding to staticmethod

[ghstack-poisoned]

awgu commented on torch/testing/_internal/distributed/_tensor/common_dtensor.py:
```diff
@@ -3,44 +3,37 @@
-from torch.utils._pytree import tree_flatten, tree_unflatten, TreeSpec
+from torch.distributed._tensor import DeviceMesh, distribute_tensor, Replicate, Shard
+from torch.distributed._tensor.placement_types import Placement
+from torch.distributed.tensor.parallel import (
```
For some reason, when I ran `lintrunner`, it could not auto format this line when it was too long, so I formatted this file locally. It looks like it made some extra format changes. I think we need to enable `lintrunner` on this file or something.

awgu commented on torch/testing/_internal/distributed/_tensor/common_dtensor.py:
```diff
@@ -140,27 +139,31 @@ def __init__(self, dim, hidden_dim, dropout_p):
         assert args.vocab_size is not None
         assert args.max_seq_len is not None
-        self.max_seq_len = args.max_seq_len
+        self.model_args = args
```
I saved the `ModelArgs` as attribute to avoid having to pass in the `ModelArgs` to `parallelize()`. Hopefully this is okay.

wanchaol replied:
not sure if there's any implication when we turn on torch.compile on 2D e2e tests 🤔 

I guess if there's issue pop up in the future, we can always refactor to pass args to parallelize directly

awgu replied:
To be safe, let me leave `self.max_seq_len` then.

wanchaol reviewed (approved):
sgtm!

Commit:
Update on "[DTensor] Moved `Transformer` sharding to staticmethod"


To support FSDP + TP/SP unit tests, let us factor out the canonical TP/SP sharding of `Transformer` to a staticmethod that can be called by other unit tests.


Test Plan:
```
pytest test/distributed/tensor/parallel/test_tp_examples.py -k test_transformer_training
```

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang

[ghstack-poisoned]

yifuwang reviewed (approved):
Looks great!

awgu commented:
@pytorchbot merge

awgu commented:
Failure in [pull / linux-focal-cuda12.1-py3.10-gcc9 / test (default, 5, 5, linux.4xlarge.nvidia.gpu)](https://github.com/pytorch/pytorch/actions/runs/8240668426/job/22537096811) looks unrelated:
```
2024-03-12T00:20:59.2140727Z FAILED [0.0008s] test_ops_fwd_gradients.py::TestFwdGradientsCUDA::test_fn_fwgrad_bwgrad_linalg_lu_factor_cuda_complex128 - RuntimeError: CUDA error: an illegal memory access was encountered
2024-03-12T00:20:59.2141431Z CUDA kernel errors might be asynchronously reported at some other API call, so the stacktrace below might be incorrect.
2024-03-12T00:20:59.2141704Z For debugging consider passing CUDA_LAUNCH_BLOCKING=1.
```

awgu commented:
@pytorchbot merge -i
